File: code/ex96.java
// Dynamic Programming

class Solution {
    public int numTrees(int n) {      
        // Init dp array
        int dp[] = new int[n + 1];  
        dp[0] = 1;  
        
        for (int i = 1; i <= n; i++)
            for (int j = 0; j < i; j++)
                dp[i] += (dp[j - 0] * dp[i - 1 - j]);

        
        return dp[n];
    }
}

File: code/ex96.cpp
class Solution {
public:
    int numTrees(int n) {
        vector<int> dp(n + 1);
        dp[0] = 1;    dp[1] = 1;
        
        for (int i = 2; i <= n; i++) {
            for (int left = 0; left < i; left++) {
                int right = i - left - 1;
                dp[i] += dp[left] * dp[right];
            } 
        }
        
        return dp[n];
    }
};